### PR TITLE
Update client.go

### PR DESCRIPTION
**Pull request by Xeonid**

### CODE REVIEW

File: bugz/client.go [hunk 1]
```diff
@@ -12,8 +12,8 @@ import (
 )
 
 type BugzClient struct {
-	token string
-	url   string
+	Token string
```
Comment (ngortheone) [inline bugz/client.go:15]: Why do you need to make token public?

File: bugz/client.go [hunk 2]
```diff
@@ -23,45 +23,65 @@ type BugzLoginResponse struct {
...
-		CheckFatal("NewBugzClient loginResponse.Token", err)
+		CheckFatal("NewBugzClient loginResponse.Token", fmt.Errorf("login token is empty"))
 	}
-	bugz.token = loginResponse.Token
+	bugz.Token = loginResponse.Token
 
 	return bugz
 }
+
+func (bc *BugzClient) GetToken() string {
```
Comment (ngortheone) [inline bugz/client.go:64]: Why do you need this method?

File: bugz/client.go [hunk 3]
```diff
@@ -23,45 +23,65 @@ type BugzLoginResponse struct {
...
+	return bc.Token
+}
+
+func (bc *BugzClient) DownloadBugs() ([]Bug, error) {
```
Comment (ngortheone) [inline bugz/client.go:68]: This method does not download **all** bugs.

**Review by ngortheone — CHANGES_REQUESTED** [review]

**Review by ngortheone — CHANGES_REQUESTED** [review]

File: bugz/client.go (hunk 2 above)
Comment (Xeonid) [reply bugz/client.go]: ok

File: main.go [hunk 4]
```diff
@@ -8,42 +8,89 @@ import (
...
-}
-func showBugs() {
-	fmt.Println("Showing bugs")
-}
 
+	func showBugs() {
+		fmt.Println("showing bugs")
+	}
+
+	/*func getBugsFromBugzilla(pullConcurrencyLevel int, chanBug chan bugz.Bug) {
```
[the remark this answers is not in the document]
Comment (Xeonid) [reply main.go]: ok

File: main.go [hunk 5]
```diff
@@ -8,42 +8,89 @@ import (
...
-	fmt.Println("Listing bugs")
-}
-func downloadUsers() {
-	fmt.Println("Downloading users")
-}
-func showBugs() {
-	fmt.Println("Showing bugs")
-}
 
+	func showBugs() {
```
[the remark this answers is not in the document]
Comment (Xeonid) [reply main.go]: ok

File: main.go [hunk 6]
```diff
@@ -8,42 +8,89 @@ import (
...
-		printHelp()
-	case "bugzilla-list-bugs":
-		listBugs()
-	case "bugzilla-download-users":
-		downloadUsers()
-	case "bugzilla-show-bugs":
-		showBugs()
-	default:
-		fmt.Println("Use help")
+	func listBugs() {
```
[the remark this answers is not in the document]
Comment (Xeonid) [reply main.go:40]: ok

File: bugz/getToken.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+)
+
+func GetToken() (string, error) {
```
[the remark this answers is not in the document]
Comment (Xeonid) [reply bugz/getToken.go]: ok

File: bugz/getIDS.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func getIDS() (map[int]User, error) {
```
[the remark this answers is not in the document]
Comment (Xeonid) [reply bugz/getIDS.go]: ok

File: bugz/client.go [hunk 9]
```diff
@@ -23,45 +22,49 @@ type BugzLoginResponse struct {
...
-		CheckFatal("NewBugzClient loginResponse.Token", err)
+		CheckFatal("NewBugzClient loginResponse.Token", fmt.Errorf("login token is empty"))
 	}
-	bugz.token = loginResponse.Token
+	bugz.Token = loginResponse.Token*/
 
 	return bugz
 }
+
+func (bc *BugzClient) GetToken() (string, error) {
```
[the remark this answers is not in the document]
Comment (Xeonid) [reply bugz/client.go:63]: ok

File: bugz/client.go (hunk 1 above)
Comment (Xeonid) [reply bugz/client.go]: ok

File: bugz/getUsers.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
+package bugz
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+func DownloadBugzillaUsers() {
```
[the remark this answers is not in the document]
Comment (Xeonid) [reply bugz/getUsers.go:10]: ok

File: bugz/bugs.go [hunk 11]
```diff
@@ -7,19 +7,26 @@ import (
 	"net/http"
 	"net/url"
 	"os"
+	"path/filepath"
 )
 
-func DownloadBugs() {
+func DownloadBugzillaBugs() error {
```
[the remark this answers is not in the document]
Comment (Xeonid) [reply bugz/bugs.go:13]: ok

File: bugz/client.go [hunk 12]
```diff
@@ -23,45 +23,49 @@ type BugzLoginResponse struct {
...
 	return bugz
 }
+
+func (bc *BugzClient) GetToken() string {
+	return bc.Token
+}
+
+// DownloadAllBugs downloads all bugs from the Bugzilla API and saves them to individual JSON files.
+func (bc *BugzClient) DownloadAllBugs() error {
+	return DownloadBugzillaBugs()
```
[the remark this answers is not in the document]
Comment (Xeonid) [reply bugz/client.go]: ok

File: bugz/client.go (hunk 3 above)
Comment (Xeonid) [reply bugz/client.go]: ok

File: bugz/bugs.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (ngortheone) [reply bugz/bugs.go:13]: why does this function exist?

File: bugz/client.go [hunk 13]
```diff
@@ -23,45 +25,242 @@ type BugzLoginResponse struct {
...
 	}
-	bugz.token = loginResponse.Token
+	bugz.Token = loginResponse.Token*/
 
 	return bugz
 }
+
+// DownloadBugzillaBugs downloads all bugs from the Bugzilla API and saves them to individual JSON files.
+func (bc *BugzClient) DownloadBugzillaBugs() error {
+	apiURL := "https://bugs.freebsd.org/bugzilla/rest/bug"
```
Comment (ngortheone) [inline bugz/client.go:68]: reuse `bc.URL here`

File: bugz/getIDS.go [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+func GetIDS() (map[int]User, error) {
+	// Specify the path to the directory containing JSON files
+	directoryPath := "bugs"
+
+	fmt.Println("Get IDS from bugs")
+
+	// Get a list of all JSON files in the directory
+	files, err := os.ReadDir(directoryPath)
+	if err != nil {
+		fmt.Printf("Error reading directory: %v\n", err)
```
Comment (ngortheone) [inline bugz/getIDS.go:20]: Normally you either return the error to the caller or handle it on the spot.
Printing the error is "handing it". Here you are trying to do a bit of both.

Suggest to replace this with 
```
if err != nil {
    return nil, fmt.Errorf("Error reading directory: %w", err)
}
```

And do the printing in the calling function

File: bugz/client.go [hunk 15]
```diff
@@ -23,45 +25,242 @@ type BugzLoginResponse struct {
...
+	err := os.MkdirAll("data/users", os.ModePerm)
+	if err != nil && !os.IsExist(err) {
+		fmt.Printf("Error creating 'users' folder: %v\n", err)
+		return nil
+	}
+
+	// Extract unique IDs from bugs
+	idUserMap, err := getIDS()
+	if err != nil {
+		return nil
```
Comment (ngortheone) [inline bugz/client.go:236]: This is wrong. 

If `err !=nil` means we have a problem. But returning `nil` to the caller you say "no problems happened".

File: bugz/client.go [hunk 16]
```diff
@@ -23,45 +25,242 @@ type BugzLoginResponse struct {
...
+	}
+
+	return idUserMap
+}
+
+func (bc *BugzClient) DownloadBugzillaUsers() error {
+	// Create a 'data/users' folder if it doesn't exist
+	err := os.MkdirAll("data/users", os.ModePerm)
+	if err != nil && !os.IsExist(err) {
+		fmt.Printf("Error creating 'users' folder: %v\n", err)
```
Comment (ngortheone) [inline bugz/client.go:229]: Returning nil to the caller you say "no problems happened".

replace this with
```
return fmt.Errorf("Error creating 'users' folder: %w", err)
```

File: bugz/client.go [hunk 17]
```diff
@@ -23,45 +25,242 @@ type BugzLoginResponse struct {
...
+func getIDS() (map[int]User, error) {
+	// Specify the path to the directory containing JSON files
+	directoryPath := "bugs"
+
+	fmt.Println("Get IDS from bugs")
+
+	// Get a list of all JSON files in the directory
+	files, err := os.ReadDir(directoryPath)
+	if err != nil {
+		fmt.Printf("Error reading directory: %v\n", err)
```
Comment (ngortheone) [inline bugz/client.go:162]: Same here. See my other comments about return values

File: bugz/client.go [hunk 18]
```diff
@@ -23,45 +25,242 @@ type BugzLoginResponse struct {
...
+		if err != nil {
+			return fmt.Errorf("Error decoding JSON: %v\n", err)
+		}
+
+		// Iterate over bugs and write to individual files
+		for _, bug := range bugsResponse["bugs"] {
+			filename := filepath.Join("bugs", fmt.Sprintf("bug_%d.json", bug.ID))
+			err := writeToFile(filename, bug)
+			if err != nil {
+				return fmt.Errorf("Error writing to file %s: %v\n", filename, err)
```
Comment (ngortheone) [inline bugz/client.go:115]: No need to add explicit newline '\n' in Errorf

File: bugz/client.go [hunk 19]
```diff
@@ -23,45 +25,242 @@ type BugzLoginResponse struct {
...
+		if err != nil {
+			return fmt.Errorf("Error reading response body from %s: %v\n", fullURL, err)
+		}
+		response.Body.Close()
+
+		// Process the JSON data
+		var bugsResponse map[string][]Bug
+		err = json.Unmarshal(body, &bugsResponse)
+		if err != nil {
+			return fmt.Errorf("Error decoding JSON: %v\n", err)
```
Comment (ngortheone) [inline bugz/client.go:107]: No need to add explicit newline '\n' in Errorf

File: bugz/client.go [hunk 20]
```diff
@@ -23,45 +25,242 @@ type BugzLoginResponse struct {
...
+		// Make a GET request to the API
+		response, err := http.Get(fullURL)
+		if err != nil {
+			return fmt.Errorf("Error making GET request to %s: %v\n", fullURL, err)
+		}
+
+		// Read the response body
+		body, err := io.ReadAll(response.Body)
+		if err != nil {
+			return fmt.Errorf("Error reading response body from %s: %v\n", fullURL, err)
```
Comment (ngortheone) [inline bugz/client.go:99]: No need to add explicit newline '\n' in Errorf

File: bugz/client.go [hunk 21]
```diff
@@ -23,45 +25,242 @@ type BugzLoginResponse struct {
...
+		params.Set("limit", fmt.Sprint(pageSize))
+		params.Set("offset", fmt.Sprint((pageNumber)*pageSize))
+
+		// Construct the full URL with query parameters
+		fullURL := apiURL + "?" + params.Encode()
+
+		// Make a GET request to the API
+		response, err := http.Get(fullURL)
+		if err != nil {
+			return fmt.Errorf("Error making GET request to %s: %v\n", fullURL, err)
```
Comment (ngortheone) [inline bugz/client.go:93]: No need to add explicit newline '\n' in Errorf

File: bugz/client.go [hunk 22]
```diff
@@ -23,45 +25,242 @@ type BugzLoginResponse struct {
...
 }
+
+// DownloadBugzillaBugs downloads all bugs from the Bugzilla API and saves them to individual JSON files.
+func (bc *BugzClient) DownloadBugzillaBugs() error {
+	apiURL := "https://bugs.freebsd.org/bugzilla/rest/bug"
+
+	// Create a 'bugs' folder if it doesn't exist
+	err := os.MkdirAll("bugs", os.ModePerm)
+	if err != nil && !os.IsExist(err) {
+		return fmt.Errorf("Error creating 'users' folder: %v\n", err)
```
Comment (ngortheone) [inline bugz/client.go:73]: No need to add explicit newline '\n' in Errorf

**Review by ngortheone — CHANGES_REQUESTED** [review]

File: bugz/client.go [hunk 23]
```diff
@@ -23,45 +25,239 @@ type BugzLoginResponse struct {
...
-	if !ok {
-		panic("BUGZILLA_LOGIN is not set")
-	}
-	password, ok := os.LookupEnv("BUGZILLA_PASSWORD")
-	if !ok {
-		panic("BUGZILLA_PASSWORD is not set")
-	}
-	var loginResponse BugzLoginResponse
+func NewBugzClient() *BugzClient {
+	//login := os.Getenv("BUGZILLA_LOGIN") //Retrieve env var values and check if they are empty
```
Comment (ngortheone) [inline bugz/client.go:29]: do not commit commented out code

**Review by ngortheone — CHANGES_REQUESTED** [review]

File: bugz/client.go [hunk 24]
```diff
@@ -23,45 +25,212 @@ type BugzLoginResponse struct {
...
+		params := url.Values{}
+		params.Set("api_key", "val")
+		params.Set("limit", fmt.Sprint(pageSize))
+		params.Set("offset", fmt.Sprint((pageNumber)*pageSize))
+
+		// Construct the full URL with query parameters
+		fullURL := apiURL + "?" + params.Encode()
+
+		// Make a GET request to the API
+		response, err := http.Get(fullURL)
```
Comment (ngortheone) [inline bugz/client.go:63]: why not `bc.http.Get(...` ? Use it since we have created it in the struct.

File: bugz/client.go [hunk 25]
```diff
@@ -23,45 +25,212 @@ type BugzLoginResponse struct {
...
-			// Timeout: time.Second * 10,
-		},
+	// Specify the pagination parameters
+	pageSize := 1000
+	pageNumber := 0
+
+	for {
+		// Create query parameters
+		params := url.Values{}
+		params.Set("api_key", "val")
```
Comment (ngortheone) [inline bugz/client.go:55]: I think here you are supposed to use `bc.token`

File: bugz/client.go [hunk 26]
```diff
@@ -23,45 +25,212 @@ type BugzLoginResponse struct {
 	Token string `json:"token"`
 }
 
-func NewBugzClient() BugzClient {
-	login, ok := os.LookupEnv("BUGZILLA_LOGIN")
-	if !ok {
-		panic("BUGZILLA_LOGIN is not set")
+func NewBugzClient() *BugzClient {
+
```
Comment (ngortheone) [inline bugz/client.go:29]: In this function you are supposed to login and use `BugzLoginResponse` to get a token and save token to the struct,

File: bugz/client.go [hunk 27]
```diff
@@ -23,45 +25,212 @@ type BugzLoginResponse struct {
...
+		}
+
+		// Move to the next page
+		pageNumber++
 	}
-	formData := url.Values{}
-	formData.Set("login", login)
-	formData.Set("password", password)
 
-	response, err := bugz.http.Get(bugz.url + "/login?" + formData.Encode())
```
Comment (ngortheone) [inline bugz/client.go:48]: You have deleted the code that does the login, and I don't see an alternative.
How does BugzlillaClient authenticate?

**Review by ngortheone — CHANGES_REQUESTED** [review]

File: main.go [hunk 28]
```diff
@@ -8,42 +8,46 @@ import (
...
+			fmt.Printf("Error downloading users: %v\n", err)
+		}
 	case "bugzilla-show-bugs":
-		showBugs()
+		if err := bc.ShowBugs(); err != nil {
+			fmt.Printf("error showing bugs: %v\n", err)
+		}
 	default:
-		fmt.Println("Use help")
+		fmt.Println("use help")
```
Comment (ngortheone) [inline main.go:41]: sidenote: you can say "invalid command" and call `printHelp()` here.

File: bugz/client.go [hunk 29]
```diff
@@ -23,45 +23,241 @@ type BugzLoginResponse struct {
...
+		http:  &http.Client{},
 	}
+
 	formData := url.Values{}
 	formData.Set("login", login)
 	formData.Set("password", password)
 
-	response, err := bugz.http.Get(bugz.url + "/login?" + formData.Encode())
-	CheckFatal("NewBugzClient bugz.http.Get", err)
+	response, _ := bc.http.Get(bc.URL + "/login?" + formData.Encode())
```
Comment (ngortheone) [inline bugz/client.go:43]: do not ignore the error here

**Review by ngortheone — CHANGES_REQUESTED** [review]

**Review by ngortheone — APPROVED** [review]

**Pull request merged**